src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -0,0 +1,436 @@
+            _api._setCallbackHandlerDelegate(
+                _nativeObjPtr,
+                _connectionDelegate,
+                GCHandle.ToIntPtr(_handle));
+        }
+
+        public Task ShutdownAsync(
+            QUIC_CONNECTION_SHUTDOWN_FLAG Flags,
+            ushort ErrorCode)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            uint status = _api._connectionShutdownDelegate(
+                _nativeObjPtr,
+                (uint)Flags,
+                ErrorCode);
+            MsQuicStatusException.ThrowIfFailed(status);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+            return _shutdownTcs.Task;
+        }
+
+        internal static uint NativeCallbackHandler(
+            IntPtr connection,
+            IntPtr context,
+            ref ConnectionEvent connectionEventStruct)
+        {
+            var handle = GCHandle.FromIntPtr(context);
+            var quicConnection = (MsQuicConnection)handle.Target;
+            // TODO quicConnection can be null
+            return quicConnection.HandleEvent(ref connectionEventStruct);
+        }
+
+        public override void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        ~MsQuicConnection()
+        {
+            Dispose(false);
+        }
+
+        private void Dispose(bool disposing)
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            if (_nativeObjPtr != IntPtr.Zero)
+            {
+                ShutdownAsync(QUIC_CONNECTION_SHUTDOWN_FLAG.NONE, 0).GetAwaiter().GetResult();
+                _api._connectionCloseDelegate?.Invoke(_nativeObjPtr);
+            }
+
+            _nativeObjPtr = IntPtr.Zero;
+            _api = null;
+
+            _handle.Free();
+            _disposed = true;
+        }
+
+        private unsafe void SetUshortParamter(QUIC_PARAM_CONN param, ushort count)
+        {
+            QuicBuffer buffer = new QuicBuffer()
+            {
+                Length = sizeof(ushort),
+                Buffer = (byte*)&count
+            };
+            SetParam(param, buffer);
+        }
+
+        private void SetParam(
+            QUIC_PARAM_CONN param,
+            QuicBuffer buf)
+        {
+            MsQuicStatusException.ThrowIfFailed(_api.UnsafeSetParam(
+                _nativeObjPtr,
+                (uint)QUIC_PARAM_LEVEL.CONNECTION,
+                (uint)param,
+                buf));
+        }
+
+        internal override void Close()
+        {
+            Dispose(false);
+        }
+
+        public override async ValueTask DisposeAsync()
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            if (_nativeObjPtr != IntPtr.Zero)
+            {
+                await ShutdownAsync(QUIC_CONNECTION_SHUTDOWN_FLAG.NONE, 0).ConfigureAwait(false);
```

Comment:
Is this a graceful shutdown?

Reply:
@nibanks is there any difference between calling Connection Close and Shutdown Silent + Close? Will Shutdown Silent trigger an event for Shutdown Completed?

Reply:
Connection close will essentially do a silent shutdown implicitly if not already done.